src/oic/oic/provider.py

```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
             except KeyError:
                 continue
         return None
+
+    # Below are LOGOUT related methods
+
+    def verify_post_logout_redirect_uri(
+        self, esreq: Message, client_id: str
+    ) -> Optional[str]:
+        """
+        Verify a post logout URI.
+
+        :param esreq: End session request
+        :param client_id: The Client ID
+        :return: The post logout URI if it was OK otherwise None
+        """
+        try:
+            redirect_uri = esreq["post_logout_redirect_uri"]
+        except KeyError:
+            logger.debug("Missing post_logout_redirect_uri parameter")
+            return None
+
+        try:
+            accepted_urls = self.cdb[client_id]["post_logout_redirect_uris"]
+            if self._verify_url(redirect_uri, accepted_urls):
+                return redirect_uri
+        except Exception as exc:
+            msg = "An error occurred while verifying redirect URI: %s"
+            logger.debug(msg, str(exc))
+
+        return None
+
+    def let_user_verify_logout(
+        self,
+        uid: str,
+        esr: Message,
+        cookie: Optional[List[Tuple[str, str]]],
+        redirect_uri: Optional[str],
+    ) -> Response:
+        """
+        Show a page to the user, that asks whether logout should be performed.
+
+        :param uid: User ID
+        :param esr: EndSessionRequest instance
+        :param cookie: A cookie
+        :param redirect_uri: URL
+        :return: Response instance
+        """
+        if cookie:
+            headers = cookie
+        else:
+            headers = []
+
+        self.sdb.set_verify_logout(uid)
+
+        if redirect_uri is not None:
+            redirect = redirect_uri
+        else:
+            redirect = "/"
+        try:
+            tmp_id_token_hint = esr["id_token_hint"]
+        except KeyError:
+            tmp_id_token_hint = ""
+
+        context = {
+            "id_token_hint": tmp_id_token_hint,
+            "post_logout_redirect_uri": esr["post_logout_redirect_uri"],
+            "key": self.sdb.get_verify_logout(uid),
+            "redirect": redirect,
+            "action": "/" + EndSessionEndpoint("").etype,
+        }
+        return Response(
+            self.template_renderer("verify_logout", context), headers=headers
+        )
+
+    def _get_uid_from_cookie(
+        self, cookie: Union[str, SimpleCookie, None]
```

Comment:
This is probably `Optional[Union[str, SimpleCookie]]`?

Reply:
Sort of the same but yes. Optional is just a shorthand.